offer-manager-domain/src/main/groovy/life/qbic/business/products/create/CreateProduct.groovy

```diff
@@ -28,15 +31,36 @@ class CreateProduct implements CreateProductInput {
     @Override
     void create(Product product) {
         try {
-            //todo create new productId!!! the id is not provided from the controller. It must be generated in the use case itself
-            dataSource.store(product)
-            output.created(product)
+            Product newProduct = buildNewProduct(product)
+            dataSource.store(newProduct)
+            output.created(newProduct)
         } catch(DatabaseQueryException databaseQueryException) {
             log.error("Product creation failed", databaseQueryException)
             output.failNotification("Could not create product $product.productName with id $product.productId")
         } catch(ProductExistsException productExistsException) {
-            log.warn("Product \"$product.productName\" already existed.", productExistsException)
+            log.warn("Product \"$product.productName\" already exists.", productExistsException)
             output.foundDuplicate(product)
         }
     }
+
+    private Product buildNewProduct(Product product){
+        ProductCategory category = Converter.getCategory(product)
+        long id = getProductVersion(category)
+
+        return Converter.createProductWithVersion(category,product.productName,product.description,product.unitPrice,product.unit,id)
+    }
+
+    private long getProductVersion(ProductCategory category){
+        Optional<ProductId> id = dataSource.fetchLatestProductIdentifierVersion(category)
+
+        if(id.isPresent()){
+            ProductId productId = id.get()
+            if(productId.uniqueId > 0) return productId.uniqueId + 1
+            throw new IllegalArgumentException("An unexpected error occurred")
+        }else{
+            //no product of this type was stored yet
+            return 1
+        }
```

Comment:
Imo this belongs to the data source implementation and not in the business logic. When you save the product, assign it the next higher identifier automatically.

Reply:
Documented in this issue https://github.com/qbicsoftware/offer-manager-2-portlet/issues/395